trollsift/tests/unittests/test_parser.py

```diff
@@ -322,9 +295,76 @@ def test_greediness(self):
         self.assertEqual(exp, res_dict)
 
 
+class TestCompose:
+    """Test routines related to `compose` methods."""
+
+    @pytest.mark.parametrize('compose_function', [compose, partial_compose])
+    def test_compose(self, compose_function):
+        """Test the compose method's custom conversion options."""
+        key_vals = {"a": "this Is A-Test b_test c test"}
+
+        new_str = compose_function("{a!c}", key_vals)
+        assert new_str == "This is a-test b_test c test"
+        new_str = compose_function("{a!h}", key_vals)
+        assert new_str == "thisisatestbtestctest"
+        new_str = compose_function("{a!H}", key_vals)
+        assert new_str == "THISISATESTBTESTCTEST"
+        new_str = compose_function("{a!l}", key_vals)
+        assert new_str == "this is a-test b_test c test"
+        new_str = compose_function("{a!R}", key_vals)
+        assert new_str == "thisIsATestbtestctest"
+        new_str = compose_function("{a!t}", key_vals)
+        assert new_str == "This Is A-Test B_Test C Test"
+        new_str = compose_function("{a!u}", key_vals)
+        assert new_str == "THIS IS A-TEST B_TEST C TEST"
+        # builtin repr
+        new_str = compose_function("{a!r}", key_vals)
+        assert new_str == "'this Is A-Test b_test c test'"
+        # no formatting
+        new_str = compose_function("{a}", key_vals)
+        assert new_str == "this Is A-Test b_test c test"
+        # bad formatter
+        with pytest.raises(ValueError):
+            new_str = compose_function("{a!X}", key_vals)
+        assert new_str == "this Is A-Test b_test c test"
+
+    def test_replace_undefined_params_with_placeholders(self):
+        """Test replace_undefined_params_with_placeholders function."""
+        original_fmt = "{foo}/{bar}/{baz:%Y}/{baz:%Y%m%d_%H%M}/{baz:%Y}/{bar:d}"
+        keyvals = {"foo": "foo", "bar": 0}
+
+        new_fmt, placeholders_dict = _replace_undefined_params_with_placeholders(
+            fmt=original_fmt, keyvals=keyvals
+        )
+        baz_1 = hex(hash("{baz:%Y}"))
+        baz_2 = hex(hash("{baz:%Y%m%d_%H%M}"))
+        expected_new_fmt = "{foo}/{bar}/" + f"({baz_1})/({baz_2})/({baz_1})/" + "{bar:d}"
+        assert new_fmt == expected_new_fmt
+
+        recovered_fmt = new_fmt
+        for placeholder, value in placeholders_dict.items():
+            recovered_fmt = recovered_fmt.replace(placeholder, value)
+        assert recovered_fmt == original_fmt
```

Comment:
I think this test function should be removed, for multiple reasons:
1. This code is exercised in `test_partial_compose_repeated_vars_with_different_formatting`
2. It is a private function (not part of the API)
3. We want to test behaviour, not implementation. This test would crash if we were to implement `partial_compose` differently in the future.

Reply:
Agree, this function is already tested through partial `compose`. I've removed this test and added a few extra test cases on partial `compose`.